exercises/src/main/scala/fpinscala/laziness/Stream.scala

```diff
@@ -17,20 +18,80 @@ trait Stream[+A] {
     case Empty => None
     case Cons(h, t) => if (f(h())) Some(h()) else t().find(f)
   }
-  def take(n: Int): Stream[A] = sys.error("todo")
 
-  def drop(n: Int): Stream[A] = sys.error("todo")
+  def take(n: Int): Stream[A] =
+    this match {
+      case Cons(h, t) if (n > 0) => Cons(h, () => t().take(n - 1))
+      case _ => Empty
+    }
+
+  @annotation.tailrec
+  final def drop(n: Int): Stream[A] =
+    this match {
+      case Empty => Empty
+      case Cons(h, t) =>
+        val tForced = t()
+        if (n < 2) tForced else tForced.drop(n - 1)
+    }
 
-  def takeWhile(p: A => Boolean): Stream[A] = sys.error("todo")
+  def takeWhile(p: A => Boolean): Stream[A] =
+    foldRight(Empty: Stream[A]) { (a, b) =>
+      if (p(a)) Cons(() => a, () => b) else Empty
+    }
 
-  def forAll(p: A => Boolean): Boolean = sys.error("todo")
+  def forAll(p: A => Boolean): Boolean =
+    foldRight(true) { (a, b) => p(a) && b }
 
-  def headOption: Option[A] = sys.error("todo")
+  def headOption: Option[A] =
+    foldRight(Option.empty[A]) { (a, _) => Some(a) }
 
   // 5.7 map, filter, append, flatmap using foldRight. Part of the exercise is
   // writing your own function signatures.
 
-  def startsWith[B](s: Stream[B]): Boolean = sys.error("todo")
+  def flatMap[B](f: A => Stream[B]): Stream[B] =
+    foldRight(Empty: Stream[B]) { (a, b) =>
+      f(a).foldRight(b) { (a, b) => Cons(() => a, () => b) }
+    }
+
+  def map[B](f: A => B): Stream[B] =
+    flatMap { a => Cons(() => f(a), () => Empty) }
+
+  def filter(p: A => Boolean): Stream[A] =
+    flatMap { a => if(p(a)) Cons(() => a, () => Empty) else Empty }
+
+  def append[B >: A](s: => Stream[B]): Stream[B] =
+    foldRight(s) { (a, b) => Cons(() => a, () => b) }
+
+  def startsWith[B >: A](s: Stream[B]): Boolean =
+    zipAll(s).takeWhile(_._2 != None).forAll { os => os._1 == os._2 }
+
+  def tails: Stream[Stream[A]] =
+    unfold(this) { s =>
+      s match {
+        case Cons(h, t) => Option(s -> t())
```

Comment:
I've started used `Option` instead of `Some` everywhere to help Scala's type inference--`Option` is guaranteed to always give you a value of type `Option`, while `Some` may give you a value of type `Some`.